crates/noirc_frontend/src/parser/parser.rs

```diff
@@ -473,6 +474,17 @@ fn bool_type() -> impl NoirParser<UnresolvedType> {
     maybe_comptime().then_ignore(keyword(Keyword::Bool)).map(UnresolvedType::Bool)
 }
 
+fn string_type<P>(expr_parser: P) -> impl NoirParser<UnresolvedType>
+where
+    P: NoirParser<Expression>,
+{
+    keyword(Keyword::String)
+        .ignore_then(
+            expr_parser.delimited_by(just(Token::LeftBracket), just(Token::RightBracket)).or_not(),
```

Comment:
Ah, here was your issue. You were using `expr_parser` for arbitrary expressions rather than the standard generics or type parser.
So for syntax like `str<N> ...` the expr_parser saw `N > ...` as an expression, and indeed distinguishing between `>` as an operator and as an end-type-bracket is ambiguous. The numeric generics PR doesn't actually allow parsing arbitrary expressions yet - only on array types which use the square brackets as well.
I was wondering whether to make it a separate PR but this can be fixed if we add a separate parser to only parse a subset of expressions valid in a type context. Notably those excluding the use of `>`.